File: apps/backend/app/services/pricing.py
"""
Pricing engine — pure functions, zero I/O, zero floats.

All amounts are integer kobo throughout. Turnaround multipliers are stored as
(numerator, denominator) rational pairs so that express (1.5×) is computed as
ceiling(kobo × 3 / 2) using only integer arithmetic. Rounding is applied once
per line (not per unit, not at the total), then lines are summed.
"""

from __future__ import annotations

from dataclasses import dataclass

# Turnaround is defined on the Order model (it is a DB column enum).
# Re-exported here so callers that only touch the pricing layer don't need
# to import from models directly.
from app.models.order import Turnaround

__all__ = ["Turnaround", "PriceRuleNotFound", "OrderLineInput", "PricedLine", "OrderPricing", "price_order"]

# Rational multipliers — (numerator, denominator). Never floats.
_MULTIPLIER: dict[Turnaround, tuple[int, int]] = {
    Turnaround.regular:  (1, 1),
    Turnaround.express:  (3, 2),
    Turnaround.same_day: (3, 1),
}


class PriceRuleNotFound(Exception):
    """Raised when no PriceRule row exists for a (service_id, tier) combination."""

    def __init__(self, service_id: str, tier: int) -> None:
        super().__init__(
            f"No price rule for service '{service_id}' at tier {tier}. "
            "Add a PriceRule row before creating orders for this service."
        )
        self.service_id = service_id
        self.tier = tier


@dataclass(frozen=True)
class OrderLineInput:
    service_id: str
    piece_count: int


@dataclass(frozen=True)
class PricedLine:
    service_id: str
    piece_count: int
    unit_price_kobo: int   # price per piece before turnaround
    line_total_kobo: int   # after multiplier, after rounding


@dataclass(frozen=True)
class OrderPricing:
    lines: tuple[PricedLine, ...]
    total_kobo: int
    turnaround: Turnaround


# ---------------------------------------------------------------------------
# Internal helpers
# ---------------------------------------------------------------------------


def _ceiling_multiply(kobo: int, num: int, den: int) -> int:
    """Return ceil(kobo × num / den) using integer arithmetic only."""
    return (kobo * num + den - 1) // den


def _price_line(unit_price_kobo: int, piece_count: int, turnaround: Turnaround) -> int:
    """
    Compute one line total in kobo.

    Rounding is applied at the line level (not per-piece, not at the order total)
    so each line's displayed price matches what is stored.
    """
    num, den = _MULTIPLIER[turnaround]
    return _ceiling_multiply(unit_price_kobo * piece_count, num, den)


# ---------------------------------------------------------------------------
# Public API
# ---------------------------------------------------------------------------


def price_order(
    tier: int,
    lines: list[OrderLineInput],
    turnaround: Turnaround,
    price_rules: dict[tuple[str, int], int],
) -> OrderPricing:
    """
    Compute the full pricing breakdown for an order.

    Args:
        tier:         Customer tier (1–3).
        lines:        Ordered list of (service_id, piece_count) inputs.
        turnaround:   Delivery speed — determines the multiplier.
        price_rules:  Pre-loaded lookup mapping (service_id, tier) → unit price in kobo.
                      Build this from PriceRule rows before calling.

    Returns:
        OrderPricing with a per-line breakdown and an integer kobo total.

    Raises:
        PriceRuleNotFound: If any line's service has no rule for the given tier.
        ValueError:        If any piece_count is less than 1.
    """
    if not lines:
        return OrderPricing(lines=(), total_kobo=0, turnaround=turnaround)

    priced: list[PricedLine] = []
    for line in lines:
        if line.piece_count < 1:
            raise ValueError(
                f"piece_count must be >= 1, got {line.piece_count} for service '{line.service_id}'"
            )

        key = (line.service_id, tier)
        if key not in price_rules:
            raise PriceRuleNotFound(line.service_id, tier)

        unit_price = price_rules[key]
        line_total = _price_line(unit_price, line.piece_count, turnaround)
        priced.append(
            PricedLine(
                service_id=line.service_id,
                piece_count=line.piece_count,
                unit_price_kobo=unit_price,
                line_total_kobo=line_total,
            )
        )

    total = sum(p.line_total_kobo for p in priced)
    return OrderPricing(lines=tuple(priced), total_kobo=total, turnaround=turnaround)

File: apps/tablet/src/services/pricing.ts
/**
 * Local pricing — mirrors the server's price_order() logic.
 * All values are integer kobo. No floats ever stored or computed.
 */

export type Turnaround = 'regular' | 'express' | 'same_day'

export interface LineInput {
  serviceId: string
  pieceCount: number
}

export interface PricedLine {
  serviceId: string
  pieceCount: number
  unitPriceKobo: number
  lineTotalKobo: number
}

export interface PricingResult {
  lines: PricedLine[]
  totalKobo: number
}

// Multipliers stored as [numerator, denominator] to stay in integer kobo
const TURNAROUND_MULTIPLIER: Record<Turnaround, [number, number]> = {
  regular:  [1, 1],
  express:  [3, 2],   // 1.5×
  same_day: [3, 1],   // 3×
}

export function applyTurnaroundMultiplier(kobo: number, turnaround: Turnaround): number {
  const [n, d] = TURNAROUND_MULTIPLIER[turnaround]
  // Integer math: multiply first, then divide and truncate
  return Math.trunc((kobo * n) / d)
}

/**
 * Look up the price (kobo) for a service at a given customer tier.
 * Falls back to tier 1 if the exact tier has no rule.
 */
export function lookupPrice(
  serviceId: string,
  tier: number,
  rules: Array<{ serviceId: string; tier: number; priceKobo: number }>,
): number {
  const exact = rules.find(r => r.serviceId === serviceId && r.tier === tier)
  if (exact) return exact.priceKobo
  const fallback = rules.find(r => r.serviceId === serviceId && r.tier === 1)
  return fallback?.priceKobo ?? 0
}

export function computeOrderTotal(
  lines: LineInput[],
  tier: number,
  turnaround: Turnaround,
  rules: Array<{ serviceId: string; tier: number; priceKobo: number }>,
): PricingResult {
  const pricedLines: PricedLine[] = lines.map(l => {
    const baseUnit = lookupPrice(l.serviceId, tier, rules)
    const unitPriceKobo = applyTurnaroundMultiplier(baseUnit, turnaround)
    return {
      serviceId: l.serviceId,
      pieceCount: l.pieceCount,
      unitPriceKobo,
      lineTotalKobo: unitPriceKobo * l.pieceCount,
    }
  })
  const totalKobo = pricedLines.reduce((sum, l) => sum + l.lineTotalKobo, 0)
  return { lines: pricedLines, totalKobo }
}
